### ISSUE TITLE
How to use CHECK constraint with a value comparison ?

### ISSUE THREAD

**Issue by hollowshiroyuki** [fnc12/sqlite_orm#1170]
I'm trying to create a table for a relationship between users but I'm having trouble using a contraint to avoid the situation where I have two records : `id1 <-> id2` and `id2 <-> id1`

```cpp
namespace orm = sqlite_orm;

struct User {
    uint16_t id = 0;
    std::string name;
};

struct FriendRequest {
    size_t id;
    uint16_t from;
    uint16_t to;
};

orm::make_table("users",
    orm::make_column("id", &User::id, orm::primary_key().autoincrement()),
    orm::make_column("name", &User::name)
)

orm::make_table("friend_requests",
    orm::make_column("id", &FriendRequest::id, orm::primary_key().autoincrement()),
    orm::make_column("from", &FriendRequest::from),
    orm::make_column("to", &FriendRequest::to),
    orm::foreign_key(&FriendRequest::from)
        .references(&User::id)
        .on_delete.cascade(),
    orm::foreign_key(&FriendRequest::to)
        .references(&User::id)
        .on_delete.cascade(),
    orm::check(orm::c(&FriendRequest::from) < orm::c(&FriendRequest::to))
```

The compiler exits with this message

```
error: no match for ‘operator<’ (operand types are ‘sqlite_orm::internal::expression_t<short unsigned int Babel::Server::ORM::FriendRequest::*>’ and ‘sqlite_orm::internal::expression_t<short unsigned int Babel::Server::ORM::FriendRequest::*>’)
   39 |                 orm::check(orm::c(&FriendRequest::from) < orm::c(&FriendRequest::to))
      |                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                  |                              |
      |                                  expression_t<[...]>            expression_t<[...]>
```

I think I followed the documentation properly, [here](https://sqliteorm.com/files/SQLITE_ORM%20Tutorial%203.2.pdf#page=62), but the example seems weird to me : why isn't the first `&Product::discount` inside a `c()` like the others ?

```cpp
make_table("products",
make_column("product_id", &Product::id, primary_key()),
make_column("product_name", &Product::name),
make_column("list_price", &Product::listPrice),
make_column("discount", &Product::discount, default_value(0)),
check(c(&Product::listPrice) >= &Product::discount and
c(&Product::discount) >= 0 and c(&Product::listPrice) >= 0)));
```

Thanks for your assistance !

**Comment by trueqbit** [comment]
Hi @hollowshiroyuki
I assume you are using sqlite_orm v1.8.2?

Anyway, what you describe here could be described as an oversight on the part of sqlite_orm.

`c()` wraps an expression (usually a value of a C++ builtin type such as a member pointer) into a C++ type for which sqlite_orm provides overloaded comparison operators. Unlike the binary and arithmetic operators, there are no overloads for the case where both arguments are such a wrapped type.

I'll provide a fix for it soon.

Meanwhile you can wrap either the left or the right expression, so
```c++
orm::check(orm::c(&FriendRequest::from) < &FriendRequest::to)
```
or
```c++
orm::check(&FriendRequest::from < orm::c(&FriendRequest::to)
```
works in your case.

**Comment by hollowshiroyuki** [comment]
Hi @trueqbit
Indeed I should've mentioned I'm using `sqlite_orm/1.8.1` from conancenter.
Thanks for the clarification, when I tried with only one of them enclosed in a `c()` it didn't work either, with this message :
```
error: no match for ‘operator<’ (operand types are ‘sqlite_orm::internal::expression_t<short unsigned int Babel::Server::ORM::FriendRequest::*>’ and ‘uint16_t Babel::Server::ORM::FriendRequest::*’ {aka ‘short unsigned int Babel::Server::ORM::FriendRequest::*’})
   64 |                         orm::check(orm::c(&FriendRequest::from) < &FriendRequest::to)
      |                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
```
Can this be due to the use of the type `std::uint16_t` in the model ?

**Comment by trueqbit** [comment]
This problem was fixed with PR #1047 and made it into [v1.8.2](fnc12/sqlite_orm/releases/tag/v1.8.2).

It looks like you are using a namespace alias, so the compiler doesn't find the operators because they are in a different namespace than the type `expression_t` that `c()` returns.
You can either `using namespace sqlite_orm;` or need to fallback to the named operators like the `lesser_than(&FriendRequest::from), &FriendRequest::to)`.

Sorry for the inconvenience...

**Comment by hollowshiroyuki** [comment]
I see, I'm using this in a school project with a mandatory library version and I'm not allowed to use `using namespace` as it's considered bad practice...

I'm gonna use the named operators as suggested, thanks again ! 👋
